sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/quantiles/TDigestQuantiles.java

```diff
@@ -0,0 +1,237 @@
+   * the relative error will be at most 3/C.
+   *
+   * <p>Example of use :
+   * <pre>{@code PCollection<KV<Integer, Double>> input = ...
+   * PCollection<KV<Integer, SerializableTDigest>> sketch = input
+   *                .apply(TDigestQuantiles.perKey(1000));
+   * }</pre>
+   *
+   * @param compression     the compression factor guarantees a relative error of at most
+   *                        {@code 3 / compression} on quantiles.
+   * @param <K>             the type of the keys
+   */
+  public static <K> Combine.PerKey<K, Double, SerializableTDigest> perKey(int compression) {
+    return Combine.<K, Double, SerializableTDigest>perKey(TDigestQuantilesFn.create(compression));
+  }
+
+  /**
+   * A {@code Combine.CombineFn} that computes the {@link SerializableTDigest} structure
+   * of an {@code Iterable} of Doubles, useful as an argument to {@link Combine#globally} or
+   * {@link Combine#perKey}.
+   */
+  public static class TDigestQuantilesFn
+      extends Combine.CombineFn<Double, SerializableTDigest, SerializableTDigest> {
+
+    private final int compression;
+
+    private TDigestQuantilesFn(int compression) {
+      this.compression = compression;
+    }
+
+    public static TDigestQuantilesFn create(int compression) {
+        if (compression > 0) {
+            return new TDigestQuantilesFn(compression);
+        }
+        throw new IllegalArgumentException("Compression factor should be greater than 0.");
+    }
+
+    @Override public SerializableTDigest createAccumulator() {
+      return new SerializableTDigest(compression);
+    }
+
+    @Override public SerializableTDigest addInput(SerializableTDigest accum, Double value) {
+      accum.add(value);
+      return accum;
+    }
+
+    @Override public SerializableTDigest extractOutput(SerializableTDigest accum) {
+      return accum;
+    }
+
+    @Override public SerializableTDigest mergeAccumulators(
+        Iterable<SerializableTDigest> accumulators) {
+      return SerializableTDigest.merge(accumulators);
+    }
+
+    @Override public Coder<SerializableTDigest> getAccumulatorCoder(CoderRegistry registry,
+        Coder inputCoder) {
+      return SerializableTDigestCoder.of();
+    }
+
+    @Override public Coder<SerializableTDigest> getDefaultOutputCoder(CoderRegistry registry,
+        Coder inputCoder) {
+      return SerializableTDigestCoder.of();
+    }
+
+    @Override public SerializableTDigest defaultValue() {
+      return new SerializableTDigest(10);
+    }
+  }
+
+  /**
+   * This class is a wrapper for MergingDigest class because it is not serializable.
+   * The problem has been issued and corrected on 3.2 version of Ted Dunning's implementation :
+   * <a>https://github.com/tdunning/t-digest</a>
+   * However, this version has not been released yet so the issue is still up-to-date.
+   */
+  public static class SerializableTDigest implements Serializable {
```

Comment:
<!--thread_id:cc_133014051_t; commit:50a69e4b9213b300d0e3830e7c663d919d55c3ae; resolved:0-->
<!--section:context-quote-->
> **jkff** wrote:
> Why does it need to implement `Serializable`?

<!--section:body-->
Based on the comment, it seems as though version 3.2 was released recently and you should be able to simplify your code here:
https://mvnrepository.com/artifact/com.tdunning/t-digest/3.2